src/caffe/test/test_net.cpp

```diff
@@ -2224,4 +2323,31 @@ TYPED_TEST(NetTest, TestReshape) {
   }
 }
 
+TYPED_TEST(NetTest, TestSkipPropagateDown) {
+  // check bottom_need_backward if skip_propagat_down is false
+  this->InitSkipPropNet(false);
+  for (int layer_id = 0; layer_id < this->net_->layers().size(); ++layer_id) {
+    if (this->net_->layer_names()[layer_id] == "loss") {
+      // access to bottom_need_backward coresponding to label's blob
+      bool need_back = this->net_->bottom_need_backward()[layer_id][1];
+      // if skip_propagate_down is false, the loss layer will try to
+      // backpropagate on labels
+      CHECK_EQ(need_back, true)
+        << "bottom_need_backward should be True";
+    }
+  }
+  // check bottom_need_backward if skip_propagat_down is true
+  this->InitSkipPropNet(true);
```

Comment:
Examining the code again, I'm not sure `layer_need_backward` is properly updated, although the `bottom_need_backward` is exact for each layers. Tomorrow I'll try to reflect the changes in `layer_need_backward` too